pkg/test/extensions/binary.go

```diff
@@ -15,71 +16,151 @@ import (
+	return &info, nil
 }
 
 // ListTests returns which tests this binary advertises.  Eventually, it should take an environment struct
 // to provide to the binary so it can determine for itself which tests are relevant.
 func (b *TestBinary) ListTests(ctx context.Context) (ExtensionTestSpecs, error) {
 	var tests ExtensionTestSpecs
 	start := time.Now()
-	binName := filepath.Base(b.path)
+	binName := filepath.Base(b.binaryPath)
 
-	b.logger.Printf("Listing tests for %q", binName)
-	command := exec.Command(b.path, "list")
+	fmt.Fprintf(os.Stderr, "Listing tests for %q", binName)
+	command := exec.Command(b.binaryPath, "list", "-o", "jsonl")
 	testList, err := runWithTimeout(ctx, command, 10*time.Minute)
 	if err != nil {
-		return nil, fmt.Errorf("failed running '%s list': %w", b.path, err)
+		return nil, fmt.Errorf("failed running '%s list': %w", b.binaryPath, err)
 	}
 	buf := bytes.NewBuffer(testList)
 	for {
 		line, err := buf.ReadString('\n')
 		if err == io.EOF {
 			break
 		}
-		if !strings.HasPrefix(line, "[{") {
+		if !strings.HasPrefix(line, "{") {
 			continue
 		}
 
-		var extensionTestSpecs ExtensionTestSpecs
-		err = json.Unmarshal([]byte(line), &extensionTestSpecs)
+		extensionTestSpec := new(ExtensionTestSpec)
+		err = json.Unmarshal([]byte(line), extensionTestSpec)
 		if err != nil {
-			return nil, err
-		}
-		for i := range extensionTestSpecs {
-			extensionTestSpecs[i].Binary = b.path
+			return nil, errors.Wrapf(err, "line: %s", line)
 		}
-		tests = append(tests, extensionTestSpecs...)
+		extensionTestSpec.Binary = b
+		tests = append(tests, extensionTestSpec)
 	}
-	b.logger.Printf("Listed %d tests for %q in %v", len(tests), binName, time.Since(start))
+	fmt.Fprintf(os.Stderr, "Listed %d tests for %q in %v", len(tests), binName, time.Since(start))
 	return tests, nil
 }
 
+// RunTests executes the named tests and returns the results.
+func (b *TestBinary) RunTests(ctx context.Context, timeout time.Duration, env []string,
+	names ...string) []*ExtensionTestResult {
+	var results []*ExtensionTestResult
+	unseenTests := sets.New[string](names...)
+	binName := filepath.Base(b.binaryPath)
+
+	// Build command
+	args := []string{"run-test"}
+	for _, name := range names {
+		args = append(args, "-n", name)
+	}
+	args = append(args, "-o", "jsonl")
+	command := exec.Command(b.binaryPath, args...)
+	if len(env) == 0 {
+		env = os.Environ()
+	}
+	command.Env = env
+
+	// Run test
+	testResult, err := runWithTimeout(ctx, command, timeout)
+	buf := bytes.NewBuffer(testResult)
+	for {
+		line, err := buf.ReadString('\n')
+		if err == io.EOF {
+			break
+		}
+		if !strings.HasPrefix(line, "{") {
+			continue
+		}
+		result := new(ExtensionTestResult)
+		err = json.Unmarshal([]byte(line), &result)
+		if err != nil {
+			panic(fmt.Sprintf("test binary %q returned unmarshallable result", binName))
+		}
+		if !unseenTests.Has(result.Name) {
+			result.Result = ResultFailed
```

Comment:
I don't follow the logic of why the name not being in the `unseenTests` set means it has failed. Can you explain this?

Reply:
I added a comment explaining the logic; let me know if it makes sense

Reply:
Ah, yes it makes sense now, thanks!